File: leetcode/0078_subsets.cs
// https://leetcode.com/problems/subsets/description/

public class Solution {
    public IList<IList<int>> Subsets(int[] nums) {
        return Subsets(nums, 0).Select(e => e.ToArray()).ToArray();
    }

    public IEnumerable<IEnumerable<int>> Subsets(int[] nums, int iStart) {
        if (iStart > nums.Length - 1)
        {
            yield return Enumerable.Empty<int>();
            yield break;
        }
        
        foreach (IEnumerable<int> ss in Subsets(nums, iStart + 1))
        {
            yield return ss;
            yield return ss.Concat(new[]{nums[iStart]});
        }
    }
}

File: leetcode/0078_subsets.cpp
// https://leetcode.com/problems/subsets/description/

class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        return subsets(nums, 0);
    }
    
    vector<vector<int>> subsets(vector<int>& nums, int iStart) {
        vector<vector<int>> r;

        if (iStart >= nums.size())
        {
            r.push_back(vector<int>());
            return r;
        }
        
        for (auto&& ss : subsets(nums, iStart + 1))
        {
            r.push_back(ss);
            auto copy = ss;
            copy.push_back(nums[iStart]);
            r.push_back(copy);
        }
        
        return r;
    }
};